drivers/xen/regions.c

```diff
@@ -0,0 +1,164 @@
+	uint64_t phys_addr;
+	uint64_t size;
+};
+
+/*
+ * At least one extended region should be provided.
+ * If there is no extended regions - alternative allocation should be implemented
+ * by the driver user and driver should be disabled.
+ */
+BUILD_ASSERT(DT_NUM_REGS(XEN_HYP_NODE) >= EXTENDED_REGIONS_IDX,
+	     "No Extended regions were provided in DT");
+
+/*
+ * Add BUILD_ASSERT that rejects extended region, starting from 0x0.
+ * This is done because mem_blocks is not designed to handle buffer
+ * starting from 0 and has assert with check. In this case starting address
+ * should be shifted to page 1. That shouldn't be a problem because
+ * minimal extended region size is set to 64MB in XEN.
+ * For example for the extended region 0->0x30000000 the following
+ * regs should be provided:
+ * regs = <.... 0x0 0x1000 0x0 0x2FFFF000 ... >;
+ */
+#define EXT_BLOCK(node_id, idx)								\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),				\
+		     "Extended region starting from 0 is not supported, please skip page 0"); \
+	SYS_MEM_BLOCKS_DEFINE_STATIC_WITH_EXT_BUF(region_##idx, XEN_PAGE_SIZE,		\
+		     XEN_PFN_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)),		\
+		     (uint8_t *)DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx));
+
+DT_FOREACH_REG(XEN_HYP_NODE, EXT_BLOCK);
+
+#define EXT_REG_POINTER(node_id, idx)					\
+	{								\
+		.block = &region_##idx,					\
+		.phys_addr = DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),	\
+		.size = DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx),		\
+	},
+
+/*
+ * TODO: For now we allocate buffer structures for DT_NUM_REGS, when we
+ * need a range 1...DT_FOREACH_REG. Although region 0 is ignored, it still
+ * takes stack space. This requires optimization
+ */
+static struct extended_regions ext_blocks[DT_NUM_REGS(XEN_HYP_NODE)] = {
+	DT_FOREACH_REG(XEN_HYP_NODE, EXT_REG_POINTER)
+};
+
+void *xen_region_get_pages(size_t nr_pages)
+{
+	int ret, i;
+	void *ptr;
+
+	/*
+	 * Locking ranges iteration, although each call sys_mem_blocks_alloc_contiguous
+	 * has it's own syncronisation, to avoid the situation of 2 parralel calls of
+	 * get_page and put_page on full buffers when put_page clears space in block 1
+	 * when get_page has already checked block 1. This will lead us to -ENOMEM while
+	 * block 1 space was freed already
+	 */
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		ret = sys_mem_blocks_alloc_contiguous(ext_blocks[i].block, nr_pages, &ptr);
+		if (ret == 0) {
+			break;
+		} else if (ret && ret != -ENOMEM) {
+			LOG_ERR("Unable to allocate pages, ret = %d\n", ret);
+			ptr = NULL;
+		}
+	}
+
+	k_mutex_unlock(&ext_regions_lock);
+
+	return ptr;
+}
+
+void xen_region_put_pages(void *ptr, size_t nr_pages)
+{
+	int ret, i;
+	sys_mem_blocks_t *block = NULL;
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		/* We can compare ptr to phys_addr because mapping is 1:1 */
```

Comment:
oof, we had a problem just a couple days ago, when similar assumption turned to be false...